File: boj/4949/4949.cpp
#include <iostream>
#include <stack>
using namespace std;

stack<int> s;
string line;

int c(int input) {
    if (input == ')') { return '('; }
    else if (input == ']') { return '['; }
    else { return '1'; }
}

int main() {
    while (true) {
        getline(cin, line);
        if (line == ".") {
            break;
            cout << "end\n";
        }

        for (int i = 0; i < line.length(); i++) {
            if (line[i] == '(' || line[i] == '[') { s.push(line[i]); }
            else if (line[i] == ')' || line[i] == ']') {
                if (s.empty() || s.top() != c(line[i])) { s.push('-'); break; }
                else { s.pop(); }
            }
        }

        if (s.empty()) { cout << "yes\n"; }
        else { cout << "no\n"; }
        
        while (!s.empty()) { s.pop(); }
    }
    return 0;
}

File: boj/4949/4949.py
counterpart = {')': '(', ']': '['}

while True:
    stack = []
    sentence = input()
    if sentence == '.':
        break
    for letter in sentence:
        if letter in counterpart.values():
            stack.append(letter)
        elif letter in counterpart.keys() and len(stack) == 0:
            stack = [-1]
            break
        elif stack and letter in counterpart.keys():
            if stack[-1] == counterpart[letter]:
                stack.pop()
            elif stack[-1] != counterpart[letter]:
                stack = [-1]
                break

    if len(stack) != 0:
        print('no')
    else:
        print('yes')
